### PR TITLE
Ensure managed memory is supported in cudf.pandas.

### PR DESCRIPTION
**Pull request by bdice**
## Description
Currently, WSL users of `cudf.pandas` can try to enable UVM (managed memory) but it is not supported by the driver. This PR detects whether UVM is supported before enabling a managed memory pool or prefetching.

Closes https://github.com/rapidsai/cudf/issues/16551.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/pandas/__init__.py [hunk 1]
```diff
@@ -26,6 +28,27 @@
 }
 
 
+def _ensure_managed_memory_is_supported(rmm_mode):
```
Comment (lithomas1) [inline python/cudf/cudf/pandas/__init__.py:31]: Can you move this function into pylibcudf so that we can re-use it for cudf-polars as well?

**Review by lithomas1 — COMMENTED** [review]
Thanks for the speedy PR!

File: python/cudf/cudf/pandas/__init__.py (hunk 1 above)
Comment (bdice) [reply python/cudf/cudf/pandas/__init__.py]: Sure! Done in 12f0973.

**Review by lithomas1 — APPROVED** [review]

File: python/cudf/cudf/_lib/pylibcudf/utils.pyx [hunk 2]
```diff
@@ -34,3 +36,22 @@ cdef vector[reference_wrapper[const scalar]] _as_vector(list source):
         c_scalars.push_back(
             reference_wrapper[constscalar](dereference((<Scalar?>slr).c_obj)))
     return c_scalars
+
+
+def _is_concurrent_managed_access_supported():
+    """Check the availability of concurrent managed access (UVM)."""
+    # This checks the availability of UVM.
+    # Note that WSL does not support managed memory.
```
Comment (bdice) [inline python/cudf/cudf/_lib/pylibcudf/utils.pyx:44]:
```suggestion
    """Check the availability of concurrent managed access (UVM).
    
    Note that WSL2 does not support managed memory.
    """
```

File: python/cudf/cudf/pandas/__init__.py [hunk 3]
```diff
@@ -40,7 +40,20 @@ def install():
...
+        pylibcudf.utils._is_concurrent_managed_access_supported()
+    )
+    default_rmm_mode = (
+        "managed_pool" if managed_memory_is_supported else "pool"
+    )
+    rmm_mode = os.getenv("CUDF_PANDAS_RMM_MODE", default_rmm_mode)
+
+    if "managed" in rmm_mode and not managed_memory_is_supported:
+        raise ValueError(
+            f"Managed memory is unsupported on this system, so the requested {rmm_mode=} is not allowed."
```
Comment (bdice) [inline python/cudf/cudf/pandas/__init__.py:54]:
```suggestion
            f"Managed memory is not supported on this system, so the requested {rmm_mode=} is invalid."
```

File: docs/cudf/source/cudf_pandas/how-it-works.md [hunk 4]
```diff
@@ -44,11 +44,17 @@ allocation may be a bottleneck depending on the workload. Managed memory
 enables oversubscribing GPU memory. This allows cudf.pandas to process
 data larger than GPU memory in many cases, without CPU (Pandas) fallback.
 
+```{note}
+CUDA Unified Memory is not supported on Windows Subsystem for Linux (WSL2), so
+`cudf.pandas` uses a non-managed pool allocator. In WSL2, `cudf.pandas` is
+limited to the memory size of the GPU it is running on.
```
Comment (vyasr) [inline docs/cudf/source/cudf_pandas/how-it-works.md:50]:
```suggestion
CUDA Managed Memory on Windows, and more specifically Windows Subsystem for
Linux (WSL2), [does not support oversubscription](
https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#system-requirements-for-unified-memory),
only unified addressing. Furthermore, managed memory on WSL2 has undesirable
performance characteristics. Therefore, `cudf.pandas` uses a non-managed pool
allocator on WSL2, so `cudf.pandas` is limited to the physical size of GPU memory.
```

**Review by vyasr — APPROVED** [review]
One suggestion to improve the comment, approving assuming that you agree with the change.

**Review by raydouglass — APPROVED** [review]

**Pull request merged**